File: Python/programmers/etc/level2/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    n = len(dungeons)
    for case in permutations(dungeons, n):
        tmp = k
        cnt = 0
        for c in case:
            if tmp < c[0]:
                continue
            tmp -= c[1]
            cnt += 1
        answer = max(answer, cnt)
        
    return answer


k = 80
dungeons = [[80,20],[50,40],[30,10]]
print(solution(k, dungeons)) # 3

File: Java/programmers/etc/level2/피로도.java
package CodeTest.Java.programmers.etc.level2;

public class 피로도 {
    static boolean check[];
    static int answer = 0;

    public int solution(int k, int[][] dungeons) {
        check = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
        return answer;
    }

    private static void dfs(int tired, int[][] dungeons, int cnt) {
        for (int i=0; i<dungeons.length; i++) {
            if (!check[i] && dungeons[i][0] <= tired) {
                check[i] = true;
                dfs(tired-dungeons[i][1], dungeons, cnt+1);
                check[i] = false;
            }
        }
        answer = Math.max(answer, cnt);
    }
}
